### PR TITLE
ICE server config refresh stability

### PR DESCRIPTION
**Pull request by MushMal**
Triggering error callback on failed ICE refresh
Properly handle expired credentials in all states
Enhanced the samples to teardown and re-create the signaling client on error
Multiple bug fixes
Added direct fault injection and control in the tests

*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by zwei333** [comment]
Retry usually is the last resort.  Maybe we should try other things first?
Right now the failure happens every 20~30 mins. So it means this retry happens pretty frequently.
Last thing is that a lot of code is in application layer (those callbacks).  Maintaining the signaling channel is SDK's job.  Application layer may not want to be involved so much.

**Comment by MushMal** [comment]
> Retry usually is the last resort. Maybe we should try other things first?
> Right now the failure happens every 20~30 mins. So it means this retry happens pretty frequently.
> Last thing is that a lot of code is in application layer (those callbacks). Maintaining the signaling channel is SDK's job. Application layer may not want to be involved so much.

Perhaps it’s not clear from the handler but the SDK does all of the retries and state transitions. The error callback is issued after the Signsling client is unable to reconfresh the ice configuration or reconnect in error when connected. These have 10 and 15 second timeouts respectively. If the signaling client can’t make any progress due to network is being off or some other major issue like auth then it’s the responsibility of the application to intervene. In the samples, I just added re-create logic but customers could choose to terminate their application or raise alarms, etc..

**Comment by zwei333** [comment]
Thanks for the explanation.

**Comment by alexw91** [comment]
# [Codecov](https://codecov.io/gh/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/pull/246?src=pr&el=h1) Report
> Merging [#246](https://codecov.io/gh/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/pull/246?src=pr&el=desc) into [master](https://codecov.io/gh/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/commit/9bb9950d16a6d584aed950445332a41aaaafe36b&el=desc) will **decrease** coverage by `0.11%`.
> The diff coverage is `74.19%`.

[Impacted file tree graph](https://codecov.io/gh/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/pull/246?src=pr&el=tree)

```diff
@@            Coverage Diff             @@
##           master     #246      +/-   ##
==========================================
- Coverage   87.03%   86.91%   -0.12%     
==========================================
  Files          33       33              
  Lines        7837     7935      +98     
==========================================
+ Hits         6821     6897      +76     
- Misses       1016     1038      +22     
```


| [Impacted Files](https://codecov.io/gh/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/pull/246?src=pr&el=tree) | Coverage Δ | |
|---|---|---|
| [src/source/Signaling/StateMachine.c](https://codecov.io/gh/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/pull/246/diff?src=pr&el=tree#diff-c3JjL3NvdXJjZS9TaWduYWxpbmcvU3RhdGVNYWNoaW5lLmM=) | `75.57% <0.00%> (-1.04%)` | :arrow_down: |

------

[Continue to review full report at Codecov](https://codecov.io/gh/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/pull/246?src=pr&el=continue).
> **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta)
> `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> Powered by [Codecov](https://codecov.io/gh/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/pull/246?src=pr&el=footer). Last update [9bb9950...35797a6](https://codecov.io/gh/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/pull/246?src=pr&el=lastupdated). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments).

**Comment by alexw91** [comment]
# [Codecov](https://codecov.io/gh/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/pull/246?src=pr&el=h1) Report
> Merging [#246](https://codecov.io/gh/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/pull/246?src=pr&el=desc) into [master](https://codecov.io/gh/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/commit/9bb9950d16a6d584aed950445332a41aaaafe36b&el=desc) will **decrease** coverage by `0.12%`.
> The diff coverage is `74.19%`.

[Impacted file tree graph](https://codecov.io/gh/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/pull/246?src=pr&el=tree)

```diff
@@            Coverage Diff             @@
##           master     #246      +/-   ##
==========================================
- Coverage   87.03%   86.90%   -0.13%     
==========================================
  Files          33       33              
  Lines        7837     7935      +98     
==========================================
+ Hits         6821     6896      +75     
- Misses       1016     1039      +23     
```


| [Impacted Files](https://codecov.io/gh/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/pull/246?src=pr&el=tree) | Coverage Δ | |
|---|---|---|
| [src/source/Ice/IceAgentStateMachine.c](https://codecov.io/gh/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/pull/246/diff?src=pr&el=tree#diff-c3JjL3NvdXJjZS9JY2UvSWNlQWdlbnRTdGF0ZU1hY2hpbmUuYw==) | `83.04% <0.00%> (-0.30%)` | :arrow_down: |
| [src/source/Signaling/StateMachine.c](https://codecov.io/gh/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/pull/246/diff?src=pr&el=tree#diff-c3JjL3NvdXJjZS9TaWduYWxpbmcvU3RhdGVNYWNoaW5lLmM=) | `75.57% <0.00%> (-1.04%)` | :arrow_down: |

------

[Continue to review full report at Codecov](https://codecov.io/gh/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/pull/246?src=pr&el=continue).
> **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta)
> `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> Powered by [Codecov](https://codecov.io/gh/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/pull/246?src=pr&el=footer). Last update [9bb9950...7038ce1](https://codecov.io/gh/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/pull/246?src=pr&el=lastupdated). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments).

**Comment by Sean-Der** [comment]
@MushMal Do you know if this fixed any of the TSAN failures? Would be great to pay those down (also would be a great data point why we should fix them faster!)

**Comment by MushMal** [comment]
> @MushMal Do you know if this fixed any of the TSAN failures? Would be great to pay those down (also would be a great data point why we should fix them faster!)

@Sean-Der not yet entirely sure but will check. I am still working on understanding the LWS crash I was able to reproduce. Should be sending another review soon

**Comment by MushMal** [comment]
Merging this PR. Had been running with various fault injections for a day and it's stable.

**Comment by truthhonestmessaging** [comment]
It is kind of hard to follow because this PR title and description is hard to read and understand. If the call is being disconnected after 30 minutes, you guys are automatically retrying from your end? In my knowledge, webRTC should retry automatically on its own so you wouldn't have to do this.
